### PR TITLE
feat: add support for Proto Columns

### PR DESCRIPTION
**Pull request by harshachinta**
[DO NOT MERGE]

This PR has the changes from
1. https://github.com/googleapis/java-spanner/pull/2155
2. https://github.com/googleapis/java-spanner/pull/2212
3. https://github.com/googleapis/java-spanner/pull/2277

### CODE REVIEW

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractResultSet.java [hunk 1]
```diff
@@ -699,6 +717,30 @@ public boolean isNull(int columnIndex) {
...
+          "Proto message may not be null. Use MyProtoClass.getDefaultInstance() as a parameter value.");
+      try {
+        return (T)
+            message
+                .toBuilder()
+                .mergeFrom(((LazyByteArray) rowData.get(columnIndex)).getByteArray().toByteArray())
+                .build();
+      } catch (InvalidProtocolBufferException e) {
+        throw SpannerExceptionFactory.asSpannerException(e);
+      }
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractResultSet.java:733]: I think that this would be more efficient, as we would then not copy the byte array around as many times. Instead, it just creates a stream of bytes from the `base64` string that can be decoded directly.

```suggestion
      try {
        return (T)
            message
                .toBuilder()
                .mergeFrom(
                    Base64.getDecoder()
                        .wrap(
                            CharSource.wrap(((LazyByteArray) rowData.get(columnIndex)).base64String)
                                .asByteSource(StandardCharsets.UTF_8)
                                .openStream()))
                .build();
      } catch (IOException ioException) {
        throw SpannerExceptionFactory.asSpannerException(ioException);
      }
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractResultSet.java [hunk 2]
```diff
@@ -891,6 +943,56 @@ protected List<String> getJsonListInternal(int columnIndex) {
...
+        List<T> protoMessagesList = new ArrayList<>(bytesArray.size());
+        for (LazyByteArray protoMessageBytes : bytesArray) {
+          if (protoMessageBytes == null) {
+            protoMessagesList.add(null);
+          } else {
+            protoMessagesList.add(
+                (T)
+                    message
+                        .toBuilder()
+                        .mergeFrom(protoMessageBytes.getByteArray().toByteArray())
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractResultSet.java:966]: Here also: I think we should apply the same optimization as mentioned above

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Type.java [hunk 3]
```diff
@@ -127,6 +127,24 @@ public static Type pgJsonb() {
     return TYPE_PG_JSONB;
   }
 
+  /**
+   * To get the descriptor for the {@code PROTO} type.
+   *
+   * @param protoTypeFqn Proto full name
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Type.java:133]: Could we add an example for the type of string that we expect here?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java [hunk 4]
```diff
@@ -450,6 +536,85 @@ public static Value pgJsonbArray(@Nullable Iterable<String> v) {
...
+      @Nullable Iterable<AbstractMessage> v, Descriptor descriptor) {
+    if (v == null) {
+      return new ProtoMessageArrayImpl(true, null, descriptor.getFullName());
+    }
+
+    List<ByteArray> serializedArray = new ArrayList<>();
+    v.forEach(
+        (message) -> {
+          if (message != null) {
+            serializedArray.add(ByteArray.copyFrom(message.toByteArray()));
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java:557]: nit: It would have been great if we could have done this conversion without copying the byte array twice (`message.toByteArray()` creates a `new byte[]`, and the `copyFrom` method then makes a copy of that again).

I had a quick look, and could not find anything, so feel free to ignore.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java [hunk 5]
```diff
@@ -1463,6 +1677,19 @@ public ByteArray getBytes() {
...
     }
 
+    @Override
+    public <T extends AbstractMessage> T getProtoMessage(T m) {
+      Preconditions.checkNotNull(
+          m,
+          "Proto message may not be null. Use MyProtoClass.getDefaultInstance() as a parameter value.");
+      checkNotNull();
+      try {
+        return (T) m.toBuilder().mergeFrom(value.getByteArray().toByteArray()).build();
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java:1687]: Here also: Any possibility that we can do this conversion without copying to a byte array multiple times?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java [hunk 6]
```diff
@@ -1480,6 +1707,75 @@ void valueToString(StringBuilder b) {
...
+    }
+
+    @Override
+    public <T extends AbstractMessage> T getProtoMessage(T m) {
+      Preconditions.checkNotNull(
+          m,
+          "Proto message may not be null. Use MyProtoClass.getDefaultInstance() as a parameter value.");
+      checkNotNull();
+      try {
+        return (T) m.toBuilder().mergeFrom(value.toByteArray()).build();
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java:1729]: Here also: Can we do a more direct conversion?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java [hunk 7]
```diff
@@ -1962,6 +2276,31 @@ public List<ByteArray> getBytesArray() {
...
+      try {
+        List<T> protoMessagesList = new ArrayList<>(value.size());
+        for (LazyByteArray protoMessageBytes : value) {
+          if (protoMessageBytes == null) {
+            protoMessagesList.add(null);
+          } else {
+            protoMessagesList.add(
+                (T)
+                    m.toBuilder()
+                        .mergeFrom(protoMessageBytes.getByteArray().toByteArray())
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java:2294]: Here also: Would it be possible to apply the same kind of optimization as above?

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/RandomResultSetGenerator.java [hunk 8]
```diff
@@ -40,73 +45,104 @@
...
+
+    // TODO(harsha): add test corresponding to this in testGetAllTypesAsString()
+    // appendProtoTypes(types, dialect);
+    Type[] typeArray = new Type[types.size()];
+    typeArray = types.toArray(typeArray);
+    return typeArray;
+  }
+
+  /** To append Proto & Enum types * */
+  private static void appendProtoTypes(List<Type> types, Dialect dialect) {
```
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/RandomResultSetGenerator.java:125]: This does not seem to be used. Could we revert the changes in this class if they are not needed/used?

File: jitpack.yml [hunk 9]
```diff
@@ -0,0 +1,2 @@
+jdk:
+  - openjdk11
```
Comment (olavloite) [inline jitpack.yml:2]: Is this something that we want to include in this PR? If so, why?

**Review by olavloite — APPROVED** [review]
Looks generally good to me, but with some requests for optimizations. Feel free to defer those to a later PR if this is urgent to get merged.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java (hunk 5 above)
Comment (harshachinta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java]: @olavloite I have applied the optimization suggested above to this case too. Please let me know if this is correct.
```
value.getByteArray().toByteArray()
```
is changed to 
```
Base64.getDecoder()
                .wrap(
                    CharSource.wrap(value.getBase64String())
                        .asByteSource(StandardCharsets.UTF_8)
                        .openStream())
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractResultSet.java (hunk 1 above)
Comment (harshachinta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractResultSet.java]: Thanks for suggesting this optimization, @olavloite. I agree with you here that we are copying the byte array multiple times which could affect performance when proto messages are big.
I have made the changes in the same PR.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractResultSet.java (hunk 2 above)
Comment (harshachinta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractResultSet.java]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Type.java (hunk 3 above)
Comment (harshachinta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/Type.java]: Added

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java (hunk 7 above)
Comment (harshachinta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java]: I have added the optimization suggested above to this.

File: jitpack.yml (hunk 9 above)
Comment (harshachinta) [reply jitpack.yml]: We don't need this file. I have removed it.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/RandomResultSetGenerator.java (hunk 8 above)
Comment (harshachinta) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/RandomResultSetGenerator.java:124]: We need this for one test. I have made the changes in this PR https://github.com/googleapis/java-spanner/pull/2845
